Inductor: fix compile error of Tensor.select + add_

blzheng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95354

Fix https://github.com/pytorch/pytorch/issues/94960

Before this pr, `auto tmp5 = decltype(tmp4)::blendv(tmp2, tmp4, tmp1);` triggered compile error because tmp1's type is  `at::vec::CPU_CAPABILITY::Vectorized<int>`, which cannot be converted to `const at::vec::CPU_CAPABILITY::Vectorized<float>&`.

```
for(long i0=0; i0<0; i0+=1)
{
    auto tmp2 = at::vec::Vectorized<float>::loadu(out_ptr2 + 16*i0);
    auto tmp0 = at::vec::Vectorized<int>(static_cast<int>(0));
    auto tmp1 = tmp0 == tmp0;
    auto tmp3 = at::vec::Vectorized<float>(static_cast<float>(1));
    auto tmp4 = tmp2 + tmp3;
    auto tmp5 = decltype(tmp4)::blendv(tmp2, tmp4, tmp1);
    tmp5.store(out_ptr0 + 16*i0);
    tmp5.store(out_ptr2 + 16*i0);
}
```
After this pr:

```
for(long i0=0; i0<0; i0+=1)
{
    auto tmp2 = at::vec::Vectorized<float>::loadu(out_ptr2 + 16*i0);
    auto tmp0 = at::vec::Vectorized<float>(static_cast<float>(0));
    auto tmp1 = tmp0 == tmp0;
    auto tmp3 = at::vec::Vectorized<float>(static_cast<float>(1));
    auto tmp4 = tmp2 + tmp3;
    auto tmp5 = decltype(tmp4)::blendv(tmp2, tmp4, tmp1);
    tmp5.store(out_ptr0 + 16*i0);
    tmp5.store(out_ptr2 + 16*i0);
}
```

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

blzheng committed:
Inductor: fix compile error of Tensor.select + add_

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -1541,8 +1541,8 @@ def select_scatter(x, src, dim: int, index: int):
-                ops.index_expr(idx[dim], torch.int32),
-                ops.index_expr(index, torch.int32),
+                ops.index_expr(idx[dim] - index, x.get_dtype()),
+                ops.constant(0, x.get_dtype()),
```
This seems bad. What if `x` is is a bool or int8 for example, we will start running into overflow issues.

blzheng committed:
Update on "Inductor: fix compile error of Tensor.select + add_"


Fix https://github.com/pytorch/pytorch/issues/94960

Before this pr, `auto tmp5 = decltype(tmp4)::blendv(tmp2, tmp4, tmp1);` triggered compile error because tmp1's type is  `at::vec::CPU_CAPABILITY::Vectorized<int>`, which cannot be converted to `const at::vec::CPU_CAPABILITY::Vectorized<float>&`.

```
for(long i0=0; i0<0; i0+=1)
{
    auto tmp2 = at::vec::Vectorized<float>::loadu(out_ptr2 + 16*i0);
    auto tmp0 = at::vec::Vectorized<int>(static_cast<int>(0));
    auto tmp1 = tmp0 == tmp0;
    auto tmp3 = at::vec::Vectorized<float>(static_cast<float>(1));
    auto tmp4 = tmp2 + tmp3;
    auto tmp5 = decltype(tmp4)::blendv(tmp2, tmp4, tmp1);
    tmp5.store(out_ptr0 + 16*i0);
    tmp5.store(out_ptr2 + 16*i0);
}
```
After this pr:

```
for(long i0=0; i0<0; i0+=1)
{
    auto tmp2 = at::vec::Vectorized<float>::loadu(out_ptr2 + 16*i0);
    auto tmp0 = at::vec::Vectorized<float>(static_cast<float>(0));
    auto tmp1 = tmp0 == tmp0;
    auto tmp3 = at::vec::Vectorized<float>(static_cast<float>(1));
    auto tmp4 = tmp2 + tmp3;
    auto tmp5 = decltype(tmp4)::blendv(tmp2, tmp4, tmp1);
    tmp5.store(out_ptr0 + 16*i0);
    tmp5.store(out_ptr2 + 16*i0);
}
```

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/lowering.py:
```diff
@@ -1541,8 +1541,8 @@ def select_scatter(x, src, dim: int, index: int):
             ops.eq(
-                ops.index_expr(idx[dim], torch.int32),
-                ops.index_expr(index, torch.int32),
+                ops.index_expr(idx[dim] - index, torch.float),
+                ops.constant(0, torch.float),
```
I don't think it is proper to address the issue here since there is nothing wrong that indices are int32 types. We should fix this from the cpp codegen when handling the vectorized bool masks represented as float vector. You can check out the `to_float_mask` in `cpp_prefix.h`.
